helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizer.java

```diff
@@ -0,0 +1,347 @@
+    Set<String> resourceFilter = new HashSet<>();
+    Map<String, String> nodeSwap = new HashMap<>(); // old instance -> new instance.
+    List<String> newInstances = new ArrayList<>();
+    List<String> instancesToRemove = new ArrayList<>();
+    Set<String> instanceFilter = new HashSet<>();
+  }
+
+  private static class ClusterState {
+    List<InstanceConfig> instanceConfigs = new ArrayList<>();
+    ClusterConfig clusterConfig;
+    List<String> resources = new ArrayList<>();
+    List<String> instances;        // cluster LiveInstance + addInstances - instancesToRemove.
+  }
+
+  // Result format: Map of resource -> partition -> instance -> state.
+  private static class AssignmentResult extends HashMap<String, Map<String, Map<String, String>>> {
+    public AssignmentResult() {
+      super();
+    }
+  }
+
+  @ResponseMetered(name = HttpConstants.WRITE_REQUEST)
+  @Timed(name = HttpConstants.WRITE_REQUEST)
+  @POST
+  public Response computePotentialAssignment(@PathParam("clusterId") String clusterId,
+      String content) {
+
+    InputFields inputFields;
+    ClusterState clusterState;
+    AssignmentResult result;
+
+    try {
+      // 1.  Try to parse the content string. If parseable, use it as a KV mapping. Otherwise, return a REASON String
+      inputFields = readInput(content);
+      // 2. Read cluster status from ZK.
+      clusterState = readClusterStateAndValidateInput(clusterId, inputFields);
+      // 3. Call rebalancer tools for each resource.
+      result = computeOptimalAssignmentForResources(inputFields, clusterState, clusterId);
+      // 4. Serialize result to JSON and return.
+      return JSONRepresentation(result);
+    } catch (InvalidParameterException ex) {
+      return badRequest(ex.getMessage());
+    } catch (JsonProcessingException e) {
+      return badRequest("Invalid input: Input can but be parsed into a KV map." + e.getMessage());
+    } catch (OutOfMemoryError e) {
+      LOG.error(
+          "OutOfMemoryError while calling AssignmentResult" + Arrays.toString(e.getStackTrace()));
+      return badRequest(
+          "Response size is too large to serialize. Please query by resources or instance filter");
+    } catch (Exception e) {
+      LOG.error("Failed to compute partition assignment:" + Arrays.toString(e.getStackTrace()));
+      return badRequest("Failed to compute partition assignment: " + e);
+    }
+  }
+
+  private InputFields readInput(String content)
+      throws InvalidParameterException, JsonProcessingException {
+
+    Map<String, Map<String, Object>> customFieldsMap;
+    InputFields inputMap = new InputFields();
+    customFieldsMap =
+        OBJECT_MAPPER.readValue(content, new TypeReference<HashMap<String, Map<String, Object>>>() {
+        });
+
+    // Content is given as a KV mapping.
+    for (Map.Entry<String, Map<String, Object>> entry : customFieldsMap.entrySet()) {
+      String key = entry.getKey();
+      switch (key) {
+        case "InstanceChange":
```

Comment:
I'd suggest to put the acceptable commands in an enum so that it is easy to track and validate. You can check the `Command` enum in `AbstractResource`